### PR TITLE
CA-199734: best-effort DP destroy and other cleanups for mirror/copy task failures

### PR DESCRIPTION
**Pull request by zli**
Alternative version of #2664.

### CODE REVIEW

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 1]
```diff
@@ -493,119 +493,117 @@ let vdi_copy_fun __context dbg vdi_map remote is_intra_pool remote_vdis so_far t
...
+
+  if mirror then
+    with_new_dp (fun new_dp ->
+      let mirror_id, remote_vdi = mirror_to_remote new_dp in
+      with_remote_vdi remote_vdi (fun remote_vdi_ref ->
+        let mirror_record = get_mirror_record ~new_dp remote_vdi remote_vdi_ref in
+        post_mirror ?mirror_id mirror_record))
+  else
+    let mirror_record = get_mirror_record vconf.location (XenAPI.VDI.get_by_uuid remote.rpc remote.session vdi_uuid) in
+    post_mirror mirror_record
```
Comment (jonludlam) [inline ocaml/xapi/xapi_vm_migrate.ml]: Do you need to call post_mirror here, or just call the continuation with the mirror record?

Comment (zli) [inline ocaml/xapi/xapi_vm_migrate.ml]: Yes, with the current code, the effect is the same.

If we want to put more common logic in post_mirror in the future, making a uniform call might make better sense. For now, indeed, the semantic is no difference.

Let me know if you'd prefer calling the continuation directly here, in which case I'll make a quick update.

Comment (jonludlam) [inline ocaml/xapi/xapi_vm_migrate.ml]: I think it'll be clearer directly calling the continuation at this point, as it's not obvious without carefully looking that the post_mirror call does nothing if mirror=false. If we want to add common logic later, we can reconsider then. Please make the update, thanks!

Comment (zli) [inline ocaml/xapi/xapi_vm_migrate.ml]: Updated, cheers!

**Comment by jonludlam** [comment]
I think this is better than #2664. :shipit:

**Pull request merged**